source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,723 @@
+                                        return rv;
+                                    }}.format(Model.returnType)).mixture;
+                        } else {
+                            return Model.withBody(q{{
+                                        return %s.init;
+                                    }}.format(Model.returnType)).mixture;
+                        }
+                    }());
+            }
+        }
+    }
+
+    GrandTour mock = new Mock!GrandTour;
+    real x;
+    int i, l;
+    mock.foo(x, i, l++) = 1;
+    assert(mock.foo(x, i, l++) == 1);
+    assert(l == 0);
+}
+
+private enum isFunctionContainer(alias T) = is(T == module) || is(T == struct)
+    || is(T == class) || is(T == interface) || is(T == union);
+
+/**
+   Return a `Function` object refracting the function overload identified by
+   `name` and `index` in `Scope`, using the value of `localName` to represent
+   `Scope`. Store `index` in the `index` property of the Function.
+
+   Params:
+   Scope = a struct, class, interface, or union
+   localName = a string mixin that represents `Scope`
+   name = the name of the function
+   index = the index of the function in the set of overloaded function called `name`
+*/
+
+Function refract(alias Scope, string localName, string name, uint index)()
+if (isFunctionContainer!Scope) {
+    return refract!(
+        __traits(getOverloads, Scope, name)[index],
+        `__traits(getOverloads, %s, "%s")[%d]`.format(localName, name, index))
+        .withIndex(index);
+}
+
+///
+unittest {
+    interface Answers {
+        int answer();
+        string answer();
+    }
+
+    alias Container = Answers;
+
+    enum answer0 = refract!(Container, "Container", "answer", 0);
+    static assert(
+        answer0.mixture ==
+        q{@system bolts.experimental.refraction.ReturnType!(__traits(getOverloads, Container, "answer")[0]) answer();});
+    static assert(answer0.index == 0);
+
+    enum answer1 = refract!(Container, "Container", "answer", 1);
+    static assert(
+        answer1.mixture ==
+        q{@system bolts.experimental.refraction.ReturnType!(__traits(getOverloads, Container, "answer")[1]) answer();});
+    static assert(answer1.index == 1);
+}
+
+private enum True(T...) = true;
+
+/**
+   Return an array of `Function` objects, refracting the functions in `Scope`
+   for which `Predicate` evaluates to `true`, using the value of `localName` to
+   represent `Scope`. `Predicate` is optional; if not specified, refract all
+   the functions. The `index` property of each `Function` is set to the index
+   of the function in its *entire* overload set (i.e. including the overloads
+   that may have been excluded by `Predicate`).
+
+   Applying this function to a module, without specifying `Predicate`, may
+   severely affect compilation time, as *all* the properties of *all* functions
+   in the module will be queried.
+
+   Params:
+   Scope = a struct, class, interface, or union
```

Comment:
WOuld name this `Aggregate`

Reply:
> WOuld name this `Aggregate`

Ah, tehre is a problem here indeed: `Scope` can be a module as well, which doesn't count as an aggregate. Adding module to the list.